ui/page/send/shared_widgets.go

```diff
@@ -0,0 +1,580 @@
+		wi.amount.setAmount(amountAtom)
+	}
+
+	if wi.exchangeRate != -1 && wi.usdExchangeSet {
+		wi.feeRateSelector.USDExchangeSet = true
+		wi.txFeeUSD = fmt.Sprintf("$%.4f", utils.CryptoToUSD(wi.exchangeRate, feeAndSize.Fee.CoinValue))
+		wi.feeRateSelector.TxFeeUSD = wi.txFeeUSD
+		wi.totalCostUSD = utils.FormatUSDBalance(wi.Printer, utils.CryptoToUSD(wi.exchangeRate, totalSendingAmount.ToCoin()))
+		wi.balanceAfterSendUSD = utils.FormatUSDBalance(wi.Printer, utils.CryptoToUSD(wi.exchangeRate, balanceAfterSend.ToCoin()))
+
+		usdAmount := utils.CryptoToUSD(wi.exchangeRate, wal.ToAmount(amountAtom).ToCoin())
+		wi.sendAmountUSD = utils.FormatUSDBalance(wi.Printer, usdAmount)
+	}
+}
+
+func (wi *sharedProperties) showBalaceAfterSend() {
+	if wi.sourceAccountSelector != nil {
+		sourceAccount := wi.sourceAccountSelector.SelectedAccount()
+		if sourceAccount.Balance == nil {
+			return
+		}
+		balanceAfterSend := sourceAccount.Balance.Spendable
+		wi.balanceAfterSend = balanceAfterSend.String()
+		wi.balanceAfterSendUSD = utils.FormatUSDBalance(wi.Printer, utils.CryptoToUSD(wi.exchangeRate, balanceAfterSend.ToCoin()))
+	}
+}
+
+func (wi *sharedProperties) amountValidationError(err string) {
+	wi.amount.setError(err)
+	wi.clearEstimates()
+}
+
+func (wi *sharedProperties) addressValidationError(err string) {
+	wi.sendDestination.setError(err)
+	wi.clearEstimates()
+}
+
+func (wi *sharedProperties) clearEstimates() {
+	wi.txFee = " - " + string(wi.selectedWallet.GetAssetType())
+	wi.feeRateSelector.TxFee = wi.txFee
+	wi.txFeeUSD = " - "
+	wi.feeRateSelector.TxFeeUSD = wi.txFeeUSD
+	wi.totalCost = " - " + string(wi.selectedWallet.GetAssetType())
+	wi.totalCostUSD = " - "
+	wi.balanceAfterSend = " - " + string(wi.selectedWallet.GetAssetType())
+	wi.balanceAfterSendUSD = " - "
+	wi.sendAmount = " - "
+	wi.sendAmountUSD = " - "
+	wi.feeRateSelector.SetFeerate(0)
+}
+
+func (wi *sharedProperties) resetFields() {
+	wi.sendDestination.clearAddressInput()
+	wi.txLabelInputEditor.Editor.SetText("")
+
+	wi.amount.resetFields()
+}
+
+func (wi *sharedProperties) isFeerateAPIApproved() bool {
+	return wi.WL.AssetsManager.IsHTTPAPIPrivacyModeOff(libUtil.FeeRateHTTPAPI)
+}
+
+func (wi *sharedProperties) handleFunc() {
+	if wi.feeRateSelector.FetchRates.Clicked() {
+		go wi.feeRateSelector.FetchFeeRate(wi.parentWindow, wi.selectedWallet)
+	}
+
+	if wi.feeRateSelector.EditRates.Clicked() {
+		wi.feeRateSelector.OnEditRateClicked(wi.selectedWallet)
+	}
+
+	wi.nextButton.SetEnabled(wi.validate())
+	wi.sendDestination.handle()
+	wi.amount.handle()
+
+	if wi.infoButton.Button.Clicked() {
+		textWithUnit := values.String(values.StrSend) + " " + string(wi.selectedWallet.GetAssetType())
+		info := modal.NewCustomModal(wi.Load).
+			Title(textWithUnit).
+			Body(values.String(values.StrSendInfo)).
+			SetPositiveButtonText(values.String(values.StrGotIt))
+		wi.parentWindow.ShowModal(info)
+	}
+
+	if wi.retryExchange.Clicked() {
+		go wi.fetchExchangeRate()
```

Comment:
This may all be just moved code, but I am concerned about the lack of concurrency handling here. For example this is setting `wi.exchangeRate` asynchronously as you check it below.